dapp_runner/descriptor/dapp.py

```diff
@@ -110,7 +119,37 @@ def __implicit_vpn(self):
                     vm.VM_CAPS_VPN
                 ]
 
+    def _resolve_dependencies(self):
+        """Resolve instantiation priorities."""
+
+        # initialize the dependency graph and add a root node
+        self._dependency_graph: networkx.DiGraph = networkx.DiGraph()
+        self._dependency_graph.add_node(DEPENDENCY_ROOT)
+
+        # for now, we only care about the order of services,
+        # later we can enhance the dependency graph to
+        # take all the other entities into consideration
+
+        for name, service in self.nodes.items():
+            if service.depends_on:
+                self._dependency_graph.add_edge(name, service.depends_on)
+            else:
+                self._dependency_graph.add_edge(DEPENDENCY_ROOT, name)
+
+        if not networkx.is_directed_acyclic_graph(self._dependency_graph):
+            raise DescriptorError(
+                "Service definition contains a circular `depends_on`."
+            )
+
+    def nodes_prioritized(self) -> Generator[Tuple[str, ServiceDescriptor], None, None]:
```

Comment:
Is there any reason to have a `Generator` here instead of just returning a `tuple`/`list`?

If there is no reason (except for 2 lines less), then I think it would be better to change this.

(Why: I remember spending multiple hours on things like:
```
nodes = descriptor.nodes_prioritized()
# Let's print nodes first to see if they are in a correct order
for node in nodes:
    print(node)
# OK, cool, now let's use them
for node in nodes:
    use_node(node)
# WTF??? What's going on?
```

)

Reply:
good point, sure, will change to a straight list

Reply:
:+1: